Sources/NIOHPACK/DynamicHeaderTable.swift

```diff
@@ -0,0 +1,166 @@
+    private var storage: HeaderTableStorage
+    
+    /// The length of the contents of the table.
+    var length: Int {
+        return self.storage.length
+    }
+    
+    /// The size to which the dynamic table may currently grow.
+    var allowedLength: Int {
+        get {
+            return self.storage.maxSize
+        }
+        set {
+            self.storage.setTableSize(to: newValue)
+        }
+    }
+    
+    var maximumTableLength: Int {
+        didSet {
+            if self.allowedLength > maximumTableLength {
+                self.allowedLength = maximumTableLength
+            }
+        }
+    }
+    
+    /// The number of items in the table.
+    var count: Int {
+        return self.storage.count
+    }
+    
+    init(allocator: ByteBufferAllocator, maximumLength: Int = DynamicHeaderTable.defaultSize) {
+        self.storage = HeaderTableStorage(allocator: allocator, maxSize: maximumLength)
+        self.maximumTableLength = maximumLength
+        self.allowedLength = maximumLength  // until we're told otherwise, this is what we assume the other side expects.
+    }
+    
+    /// Subscripts into the dynamic table alone, using a zero-based index.
+    subscript(i: Int) -> HeaderTableEntry {
+        return self.storage[i]
+    }
+    
+    func view(of index: HPACKHeaderIndex) -> ByteBufferView {
+        return self.storage.view(of: index)
+    }
+    
+    // internal for testing
+    func dumpHeaders() -> String {
+        return self.storage.dumpHeaders(offsetBy: StaticHeaderTable.count)
+    }
+    
+    // internal for testing -- clears the dynamic table
+    mutating func clear() {
+        self.storage.purge(toRelease: self.storage.length)
+    }
+    
+    /// Searches the table for a matching header, optionally with a particular value. If
+    /// a match is found, returns the index of the item and an indication whether it contained
+    /// the matching value as well.
+    ///
+    /// Invariants: If `value` is `nil`, result `containsValue` is `false`.
+    ///
+    /// - Parameters:
+    ///   - name: The name of the header for which to search.
+    ///   - value: Optional value for the header to find. Default is `nil`.
+    /// - Returns: A tuple containing the matching index and, if a value was specified as a
+    ///            parameter, an indication whether that value was also found. Returns `nil`
+    ///            if no matching header name could be located.
+    func findExistingHeader<Name: Collection, Value: Collection>(named name: Name, value: Value?) -> (index: Int, containsValue: Bool)? where Name.Element == UInt8, Value.Element == UInt8 {
+        // looking for both name and value, but can settle for just name if no value
+        // has been provided. Return the first matching name (lowest index) in that case.
+        guard let value = value else {
+            return self.storage.indices(matching: name).first.map { ($0, false) }
+        }
```

Comment:
`findExistingHeader` currently does a linear scan of the whole table even when it is only looking for the first instance of a name, disregarding the value. That's not necessary: we can exit early. We should either make `indices` produce a lazy `Collection`, or alternatively have a `firstIndex(matching:)` method that can exit early.

Reply:
After further discussion I think a lazy `Collection` or `Sequence` is the way to go here, even for the case when we scan the whole table. We don't need an array allocation here really, and if we can take that allocation off the hot path it'd be very helpful.

This means `HeaderTableStorage` will need an as-private-as-possible method `func index<Name: Collection>(matching: Name, after index: Int) -> Int?` to base that iterator off, but that's fine, I think. (Incidentally, the initial value to `after` should be `-1` ;) ).

Reply:
I believe the standard library can help us here:

```swift
return self.headers.enumerated().lazy.filter {
    self.buffer.equalCaseInsensitiveASCII(view: name, at: $1.name)
}.map { (idx, header) in idx }
```

Reply:
That's for `indices(matching:)`. `findHeaders(matching:)` is a little nicer, since it skips the `enumerated()` and thus the `map(_:)`:

```swift
return self.headers.lazy.filter { self.buffer.equalCaseInsensitiveASCII(view: name, at: $0.name) }
```

Reply:
Well, I've started it, but the Swift 4.2 compiler segfaults when building the unit tests. Sigh. 4.1 seems ok though.

Reply:
Holy crap the new compiler optimizes well. Unit tests with a release build went from running in 14 seconds to 32 when switching to the current release version of Xcode 😮

Reply:
Okay, this is all done now, and everything seems happy in shipping Xcode etc. I'll give it another run with the beta toolset and see if I can't get a decent bug report.

Reply:
Yep, still segfaults while compiling an `@_specialized` method used by the unit tests. Only when building the unit tests, though.

```
1.	While running pass #55813 SILFunctionTransform "SimplifyCFG" on SILFunction "@$S8NIOHPACK18IndexedHeaderTableV3add15headerNameBytes05valueH0yx_q_tK3NIO20ContiguousCollectionRzAgHR_s5UInt8V7ElementRtzAjKRt_r0_lF".
 for 'add(headerNameBytes:valueBytes:)' at /Users/jdovey/Projects/Github/Quatermain/swift-nio-http2/Sources/NIOHPACK/IndexedHeaderTable.swift:146:21
error: Segmentation fault: 11
```

Reply:
Filed a bug with the Swift team: [SR-8386](https://bugs.swift.org/browse/SR-8386).